### PR TITLE
Add object lock related flags

### PR DESCRIPTION
**Pull request by nitisht**
- Adds --retention-mode,--retention-duration & --legal-hold flags
in cp command

- Adds --bypass flag in rm command

Depends on https://github.com/minio/minio-go/pull/1258

### CODE REVIEW

File: cmd/client-s3.go [hunk 1]
```diff
@@ -77,6 +77,8 @@ const (
 	AmzObjectLockMode = "X-Amz-Object-Lock-Mode"
 	// AmzObjectLockRetainUntilDate sets object lock retain until date
 	AmzObjectLockRetainUntilDate = "X-Amz-Object-Lock-Retain-Until-Date"
+	// AmzObjectLock sets object lock legal hold
+	AmzObjectLock = "X-Amz-Object-Lock-Legal-Hold"
```
Comment (kannappanr) [inline cmd/client-s3.go]:
```suggestion
	AmzObjectLockLegalHold = "X-Amz-Object-Lock-Legal-Hold"
```

**Comment by kannappanr** [comment]
```
$ mc cp /etc/issue myminio/test/issue2 --legal-hold on
mc: <ERROR> Failed to copy `/etc/issue`. on unsupported legal-hold status
```
on/ON both should be supported

**Comment by kannappanr** [comment]
```
$ mc cp /etc/issue myminio/test/issue1 --retention-mode governance --retention-duration 1d
 mc: <ERROR> invalid retention mode 'governance': invalid argument.
```
lowr case values should be honored. Both `governance` and `compliance` should work

**Comment by kannappanr** [comment]
Also, `--bypassgovernance` flag should be supported. and examples should be added

**Comment by kannappanr** [comment]
```
$ mc cp /etc/issue myminio/test/issue3 --legal-hold on --json
{
 "status": "success",
 "source": "/etc/issue",
 "target": "myminio/test/issue3",
 "size": 26,
 "totalCount": 0,
 "totalSize": 0
}
{
 "status": "success",
 "total": 0,
 "transferred": 26,
 "speed": 3153.2289792415663
}

```

**Comment by kannappanr** [comment]
```
$ mc cp /etc/issue myminio/test/issue11 --retention-mode governance --retention-duration 2d --json
{
 "status": "success",
 "source": "/etc/issue",
 "target": "myminio/test/issue11",
 "size": 26,
 "totalCount": 0,
 "totalSize": 0
}
{
 "status": "success",
 "total": 0,
 "transferred": 26,
 "speed": 2072.191650310964
}

```
Why is the total count 0. The second json structure doesn't look appropriate here. Is it something that is already in cp?

**Comment by kannappanr** [comment]
Except for the CI build failure and json output (might not be related to this PR) everything looks good to me

**Comment by nitisht** [comment]
> Except for the CI build failure and json output (might not be related to this PR) everything looks good to me

CI Failure is fixed now. 

> Why is the total count 0. The second json structure doesn't look appropriate here. Is it something that is already in cp?

File: cmd/cp-main.go [hunk 2]
```diff
@@ -146,11 +162,19 @@ EXAMPLES:
   15. Copy a text file to an object storage and preserve the file system attribute as metadata.
       {{.Prompt}} {{.HelpName}} -a myobject.txt play/mybucket
 
+<<<<<<< HEAD
```
Comment (kannappanr) [inline cmd/cp-main.go:165]: Please resolve the conflicts here

File: cmd/cp-url-syntax.go [hunk 3]
```diff
@@ -55,6 +55,22 @@ func checkCopySyntax(ctx *cli.Context, encKeyDB map[string][]prefixSSEPair) {
...
 
+	if ctx.String(rdFlag) != "" && ctx.String(rmFlag) == "" {
+		fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("Both object retention flags `--%s` and `--%s` are required.\n", rdFlag, rmFlag))
+	}
+
+	if ctx.String(rdFlag) == "" && ctx.String(rmFlag) != "" {
+		fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("Both object retention flags `--%s` and `--%s` are required.\n", rdFlag, rmFlag))
+	}
+
+	if ctx.String(lhFlag) != "" && ctx.String(rmFlag) != "" {
```
Comment (kannappanr) [inline cmd/cp-url-syntax.go:66]: legal hold is independent of retention

File: cmd/cp-url-syntax.go [hunk 4]
```diff
@@ -55,6 +55,18 @@ func checkCopySyntax(ctx *cli.Context, encKeyDB map[string][]prefixSSEPair) {
...
 
+	if ctx.String(rdFlag) != "" && ctx.String(rmFlag) == "" {
+		fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("Both object retention flags `--%s` and `--%s` are required.\n", rdFlag, rmFlag))
+	}
+
+	if ctx.String(rdFlag) == "" && ctx.String(rmFlag) != "" {
+		fatalIf(errInvalidArgument().Trace(), fmt.Sprintf("Both object retention flags `--%s` and `--%s` are required.\n", rdFlag, rmFlag))
+	}
+
+	if ctx.String(lhFlag) != "" && ctx.String(rdFlag) != "" {
```
Comment (kannappanr) [inline cmd/cp-url-syntax.go:66]: this check also should be removed

**Comment by nitisht** [comment]
ping @vadmeste PTAL

File: cmd/cp-main.go [hunk 5]
```diff
@@ -73,9 +73,25 @@ var (
 			Name:  "disable-multipart",
 			Usage: "disable multipart upload feature",
 		},
+		cli.StringFlag{
+			Name:  rmFlag,
+			Usage: "retention mode to be applied on the object",
```
Comment (vadmeste) [inline cmd/cp-main.go:78]: Could you add possible values here ?

File: cmd/cp-main.go [hunk 6]
```diff
@@ -73,9 +73,25 @@ var (
...
+		},
+		cli.StringFlag{
+			Name:  rdFlag,
+			Usage: "retention duration for the object",
```
Comment (vadmeste) [inline cmd/cp-main.go:82]: Could you add (in days or in years) here ?

File: cmd/cp-main.go [hunk 7]
```diff
@@ -73,9 +73,25 @@ var (
...
+		},
+		cli.StringFlag{
+			Name:  lhFlag,
+			Usage: "apply legal hold to the copied object",
```
Comment (vadmeste) [inline cmd/cp-main.go:86]: Could you add possible values here ? (I feel like it doesn't have any value here)

File: cmd/cp-url-syntax.go [hunk 8]
```diff
@@ -55,6 +55,14 @@ func checkCopySyntax(ctx *cli.Context, encKeyDB map[string][]prefixSSEPair) {
 		}
 	}
 
+	if ctx.String(rdFlag) != "" && ctx.String(rmFlag) == "" {
```
[the remark this answers is not in the document]
Comment (nitisht) [reply cmd/cp-url-syntax.go:58]: This would mean, cp will always expect either `rdFlag` or `rmFlag` which is not correct.

**Pull request merged**